smart_simulation/ds_tools/sma_forecast.py

```diff
@@ -0,0 +1,303 @@
+    Args:
+        start_weight: Starting weight at time of forecast.
+        consumption_series: Daily consumption (true or predicted) as a Pandas Series.
+
+    Returns: Theoretical weights series.
+    """
+    consumption_schema = pas.consumption_series
+    try:
+        de.validate_data(consumption_series, consumption_schema)
+    except Exception as ex:
+        raise ex
+
+    if not isinstance(start_weight, float):
+        try:
+            averaging_window = float(start_weight)
+        except Exception as ex:
+            raise ex
+
+    weights = consumption_series.copy() * -1
+    weights.iloc[0] += start_weight
+    weights = weights.cumsum().rename("weight")
+    return weights
+
+
+def test_weight_binary(start_weight: float, consumption_series: pd.Series) -> pd.Series:
+    """
+    Calculate theoretical binary weight of scale based on starting weight and consumption (true or predicted).
+    0 ~ scale weight <= 0
+    1 ~ scale weight > 0
+    Args:
+        start_weight: Starting weight at time of forecast.
+        consumption_series: Daily consumption (true or predicted) as a Pandas Series.
+
+    Returns: Theoretical binary weights series.
+    """
+    scale_weight_positive = calculate_test_weights(start_weight, consumption_series)
+    scale_weight_positive.loc[scale_weight_positive <= 0] = 0
+    scale_weight_positive.loc[scale_weight_positive > 0] = 1
+    return scale_weight_positive
+
+
+def train_weights(weight_series: pd.Series, train_dates: pd.DatetimeIndex) -> pd.Series:
+    """
+    Get the weight_series subset for training dates and return as a Pandas Series.
+    Args:
+        weight_series: Weights as a Pandas Series.
+        train_dates: Training dates as a Pandas DataTimeIndex.
+
+    Returns: Subset of weights for training dates.
+    """
+    weight_schema = pas.weight_series
+    try:
+        de.validate_data(weight_series, weight_schema)
+    except Exception as ex:
+        raise ex
+
+    t_weights = weight_series.copy().loc[train_dates]
+    return t_weights
+
+
+def single_test(
+    training_weights: pd.Series,
+    consumption_series_train: pd.Series,
+    consumption_series_test: pd.Series,
+    sma_window: int,
+) -> dict:
+    """
+    Forecast consumption for a single train/test dataset.
+    Args:
+        training_weights: Weights for the training dates.
+        consumption_series_train: Consumption of training dates.
+        consumption_series_test: True consumption values for test dates.
+        sma_window: Size of the averaging window to use for forecasting.
+
+    Returns: Dictionary of train, test, and prediction datasets.
+            Datasets in dictionary: train_weight, train_consumption, test_weight, test_consumption, test_binary,
+                                    pred_weight, pred_consumption, pred_binary
+
+    """
+    weight_schema = pas.weight_series
+    try:
+        de.validate_data(training_weights, weight_schema)
+    except Exception as ex:
+        raise ex
```

Comment:
If you're going to raise for every exception then you don't need a `try/except` block. You'd use a try/except if you want to do something special right here to address the exception.

Reply:
Understood. @gallamine, **so just leave the validation as `de.validate_data(training_weights, weight_schema)`, and let the exceptions come from `date_eng.validate_data()`?**

Reply:
Discussed on our call. Will make appropriate changes.